File: app/src/main/java/li/klass/fhem/domain/setlist/SetList.java
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.domain.setlist;

import com.google.common.collect.Maps;

import org.apache.commons.lang3.StringUtils;

import java.io.Serializable;
import java.util.Collections;
import java.util.List;
import java.util.Map;

import static com.google.common.collect.Lists.newArrayList;
import static org.apache.commons.lang3.StringUtils.isEmpty;

public class SetList implements Serializable  {
    private Map<String, SetListValue> entries = Maps.newHashMap();

    public void parse(String text) {
        if (isEmpty(text)) return;

        text = text.trim();

        String[] parts = text.split(" ");
        for (String part : parts) {
            handlePart(part);
        }
    }

    private void handlePart(String part) {
        if (! part.contains(":")) {
            entries.put(part, SetListEmptyValue.INSTANCE);
            return;
        }

        String[] keyValue = part.split(":");
        String key = keyValue[0];
        String value = keyValue.length == 2 ? keyValue[1] : "";

        entries.put(key, handleValue(value));
    }

    private SetListValue handleValue(String value) {
        String[] parts = value.split(",");

        if (parts.length == 4 && parts[0].equals("slider")) {
            return new SetListSliderValue(parts);
        }

        return new SetListGroupValue(parts);
    }

    public List<String> getSortedKeys() {
        List<String> keys = newArrayList(entries.keySet());
        Collections.sort(keys);
        return keys;
    }


    public Map<String, SetListValue> getEntries() {
        return Collections.unmodifiableMap(entries);
    }

    public SetListValue get(String key) {
        return entries.get(key);
    }

    public boolean contains(String ... keys) {
        if (keys == null) return false;

        for (String key : keys) {
            if (! entries.containsKey(key)) {
                return false;
            }
        }
        return true;
    }

    public int size() {
        return entries.size();
    }

    @Override
    public String toString() {
        List<String> keys = getSortedKeys();

        List<String> parts = newArrayList();
        for (String key : keys) {
            SetListValue value = entries.get(key);
            String text = value.asText();
            if (text == null) {
                parts.add(key);
            } else {
                parts.add(key + ":" + text);
            }
        }

        return StringUtils.join(parts, " ");
    }
}

File: app/src/main/java/li/klass/fhem/domain/setlist/SetList.kt
/*
 * AndFHEM - Open Source Android application to control a FHEM home automation
 * server.
 *
 * Copyright (c) 2011, Matthias Klass or third-party contributors as
 * indicated by the @author tags or express copyright attribution
 * statements applied by the authors.  All third-party contributions are
 * distributed under license by Red Hat Inc.
 *
 * This copyrighted material is made available to anyone wishing to use, modify,
 * copy, or redistribute it subject to the terms and conditions of the GNU GENERAL PUBLIC LICENSE, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU GENERAL PUBLIC LICENSE
 * for more details.
 *
 * You should have received a copy of the GNU GENERAL PUBLIC LICENSE
 * along with this distribution; if not, write to:
 *   Free Software Foundation, Inc.
 *   51 Franklin Street, Fifth Floor
 *   Boston, MA  02110-1301  USA
 */

package li.klass.fhem.domain.setlist

import li.klass.fhem.domain.setlist.typeEntry.NoArgSetListEntry
import li.klass.fhem.domain.setlist.typeEntry.NotFoundSetListEntry
import org.apache.commons.lang3.StringUtils
import org.apache.commons.lang3.StringUtils.isEmpty
import java.io.Serializable

class SetList constructor(val entries: Map<String, SetListEntry>) : Serializable {

    val sortedKeys: List<String>
        get() = entries.keys.sorted()

    operator fun get(key: String, ignoreCase: Boolean = true): SetListEntry {
        val matches = entries.filter { it.key.equals(key, ignoreCase) }
        return if (matches.isEmpty()) NotFoundSetListEntry(key) else matches.iterator().next().value
    }

    fun contains(vararg keys: String): Boolean = contains(keys.toList())

    fun contains(keys: Iterable<String>): Boolean =
            keys.any { entries.containsKey(it) }

    fun existingStatesOf(toSearch: Set<String>): Set<String> =
            entries.keys.filter { toSearch.contains(it) }.toSet()

    fun size(): Int = entries.size

    fun isEmpty(): Boolean = size() == 0

    override fun toString(): String {
        val keys = sortedKeys

        return keys
                .map { entries[it] as SetListEntry }
                .map { it.asText() }
                .joinToString(separator = " ")
    }

    fun getFirstPresentStateOf(vararg states: String): String? =
            states.asSequence().firstOrNull { contains(it) }

    companion object {
        fun parse(inputText: String): SetList {
            if (isEmpty(inputText)) return SetList(emptyMap())

            val parts = inputText.trim()
                    .split(" ".toRegex()).toTypedArray()

            return SetList(parts.map { handlePart(it) }
                    .filterNotNull()
                    .toMap())
        }

        private fun handlePart(part: String): Pair<String, SetListEntry>? {
            var myPart = part
            if (myPart.matches("[^:]+:noArg$".toRegex())) {
                myPart = myPart.replace(":noArg$".toRegex(), "")
            }
            if (!myPart.contains(":")) {
                return Pair(myPart, NoArgSetListEntry(myPart))
            }

            val keyValue = myPart.split(":".toRegex(), 2).toTypedArray()

            var key: String? = StringUtils.trimToNull(keyValue[0])
            key = if (key == null) "state" else key
            val value = if (keyValue.size == 2) keyValue[1] else ""
            if (StringUtils.isEmpty(value)) return null

            return handle(key, value)?.let { Pair(key, it) }
        }

        private fun handle(key: String, value: String): SetListItem? {
            val parts = value.split(",".toRegex()).dropLastWhile { it.isEmpty() }.toList()

            val type = findType(parts)
            return type.getSetListItemFor(key, parts)
        }

        private fun findType(parts: List<String>): SetListItemType {
            for (type in SetListItemType.values()) {
                if (type.supports(parts)) {
                    return type
                }
            }
            return if (parts.isEmpty() || parts.isNotEmpty() && "colorpicker".equals(parts[0], ignoreCase = true))
                SetListItemType.NO_ARG
            else
                SetListItemType.GROUP
        }
    }
}